aws-datastore/src/androidTest/java/com/amplifyframework/datastore/storage/sqlite/SQLiteStorageAdapterInstrumentedTest.java

```diff
@@ -276,6 +276,57 @@ public void saveModelWithMaliciousInputs() throws DataStoreException {
         assertEquals(blogOwner, result.next());
     }
 
+    /**
+     * Test save with predicate. Conditional write is useful for making sure that
+     * no data is overwritten with outdated assumptions.
+     * @throws DataStoreException from possible underlying DataStore exceptions
+     */
+    @Test
+    public void saveModelWithPredicateUpdatesConditionally() throws DataStoreException {
+        final BlogOwner john = BlogOwner.builder()
+                .name("John")
+                .build();
+        final BlogOwner jane = BlogOwner.builder()
+                .name("Jane")
+                .build();
+        final BlogOwner mark = BlogOwner.builder()
+                .name("Mark")
+                .build();
+        saveModel(john);
+        saveModel(jane);
+        saveModel(mark);
+
+        // Only update John and Jane
+        final QueryPredicate predicate = BlogOwner.NAME.beginsWith("J");
+        final BlogOwner newJohn = john.copyOfBuilder()
+                .name("John Doe")
+                .build();
+        final BlogOwner newJane = jane.copyOfBuilder()
+                .name("Jane Doe")
+                .build();
+        final BlogOwner newMark = mark.copyOfBuilder()
+                .name("Mark Doe")
+                .build();
+        saveModel(newJohn, predicate);
+        saveModel(newJane, predicate);
+        saveModel(newMark, predicate); // Should not update
+
+        Iterator<BlogOwner> blogOwners = queryModel(BlogOwner.class);
+        assertNotNull(blogOwners);
+        Set<BlogOwner> actualBlogOwners = new HashSet<>();
+        while (blogOwners.hasNext()) {
+            actualBlogOwners.add(blogOwners.next());
+        }
```

Comment:
Should we update `queryModel` to perform this `Set` building, itself? So it can just return `Set<T>` directly?

Reply:
Hmm, a nice idea, but I think I'll leave it out of this PR since this contains some other refactoring changes already haha